src/main/scala/io/iohk/ethereum/blockchain/sync/BlockchainHostActor.scala

```diff
@@ -0,0 +1,94 @@
+package io.iohk.ethereum.blockchain.sync
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
+import akka.util.ByteString
+import io.iohk.ethereum.domain.{BlockHeader, Blockchain}
+import io.iohk.ethereum.network.PeerEventBusActor.PeerEvent.MessageFromPeer
+import io.iohk.ethereum.network.PeerEventBusActor.SubscriptionClassifier.MessageReceivedClassifier
+import io.iohk.ethereum.network.PeerEventBusActor.{PeerSelector, Subscribe}
+import io.iohk.ethereum.network.PeerManagerActor.PeerConfiguration
+import io.iohk.ethereum.network.p2p.{Message, MessageSerializable}
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockBodies, BlockHeaders, GetBlockBodies, GetBlockHeaders}
+import io.iohk.ethereum.network.p2p.messages.PV63.{GetNodeData, GetReceipts, NodeData, Receipts}
+import io.iohk.ethereum.network.PeerManagerActor
+
+class BlockchainHostActor(blockchain: Blockchain, peerConfiguration: PeerConfiguration,
+                          peerEventBusActor: ActorRef, peerManagerActor: ActorRef) extends Actor with ActorLogging {
+
+  private val requestMsgsCodes = Set(GetNodeData.code, GetReceipts.code, GetBlockBodies.code, GetBlockHeaders.code)
+  peerEventBusActor ! Subscribe(MessageReceivedClassifier(requestMsgsCodes, PeerSelector.AllPeers))
+
+  override def receive: Receive = {
+    case MessageFromPeer(message, peerId) =>
+      val responseOpt = handleBlockFastDownload(message) orElse handleEvmMptFastDownload(message)
+      responseOpt.foreach{ response =>
+        peerManagerActor ! PeerManagerActor.SendMessage(response, peerId)
+      }
+  }
+
+  private def handleEvmMptFastDownload(message: Message): Option[MessageSerializable] = message match {
+    case request: GetNodeData =>
+
```

Comment:
That's right, the `GetNodeData` is used both for requesting MPT nodes and EVM code (which I agree is kind of weird), I'll refactor this code to make it clearer :+1: